### ISSUE TITLE
Intermittent failure downloading plugin

### ISSUE THREAD

**Issue by mikenorgate** [hashicorp/terraform#16112]
### Terraform Version
0.10.2

Running `terraform init` occasionally fails to download plugins. It tends to happen continuously for a short period of time and then works again. Always seems to be when getting the checksums

### Crash Output
Error installing provider "null": error fetching checksums: Get https://releases.hashicorp.com/terraform-provider-null/0.1.0/terraform-provider-null_0.1.0_SHA256SUMS: read tcp 172.27.203.11:49175->151.101.17.183:443: wsarecv: An existing connection was forcibly closed by the remote host..

**Comment by kgignatyev-inspur** [comment]
happens a lot while in China 
Error installing provider "aws": Get https://releases.hashicorp.com/terraform-provider-aws/: dial tcp 151.101.77.183:443: i/o timeout.
Error installing provider "null": Get https://releases.hashicorp.com/terraform-provider-null/: dial tcp 151.101.77.183:443: i/o timeout.

**Comment by kgignatyev-inspur** [comment]
And I can reach those urls from browser without much delay

**Comment by apparentlymart** [comment]
Hi @mikenorgate, @kgignatyev-inspur,

I'm sorry that you're both having trouble reaching `releases.hashicorp.com`. It _is_ being served through a geo-distributed CDN but it's hard to tell what might cause problems from different situations. As far as we know, Terraform isn't doing anything particularly unusual from the perspective of downloading these files, so I don't know why the behavior would be significantly different than in a browser.

We recently added [the provider plugin cache feature](https://www.terraform.io/docs/configuration/providers.html#provider-plugin-cache) to help with this sort of situation by allowing each distinct plugin to be downloaded only once and cached on your local computer for future use. Of course this assumes that you're able to download each plugin _once_, but hopefully that's easier to achieve with intermittent connectivity issues than having to download the same plugin _repeatedly_.

**Comment by akitekt** [comment]
@apparentlymart We are working behind a corporate proxy, currently we pre-download it as a cache, however, they can get out of date, it will be great if the plugin repo can be supported by tools like Artifactory as a remote repo, similar to docker remote or k8s helm remote.

**Comment by remoe** [comment]
i have the same with terraform 0.11.10

```
Initializing provider plugins...

- Checking for available provider plugins on https://releases.hashicorp.com...
Error installing provider "local": Get https://releases.hashicorp.com/terraform-provider-local/: dial tcp: i/o timeout.
```

I can open it on my browser

**Comment by michalmedvecky** [comment]
+1

**Comment by DenisBY** [comment]
Still happens:

```
- Downloading plugin for provider "aws" (hashicorp/aws) 3.1.0...

Error installing provider "aws": read tcp [2a02:8070:e183:1a00::f102]:60856->[2a04:4e42:1b::439]:443: read: connection reset by peer.

Terraform analyses the configuration and state and automatically downloads
plugins for the providers used. However, when attempting to download this
plugin an unexpected error occurred.

This may be caused if for some reason Terraform is unable to reach the
plugin repository. The repository may be unreachable if access is blocked
by a firewall.

If automatic installation is not possible or desirable in your environment,
you may alternatively manually install plugins by downloading a suitable
distribution package and placing the plugin's executable file in the
following directory:
    terraform.d/plugins/linux_amd64
```

**Comment by mildwonkey** [comment]
I am going to close this issue due to inactivity. The underlying code has been entirely rewritten since this issue was opened.


If there is still a question, I recommend the [the community forum](https://discuss.hashicorp.com/c/terraform-core), where there are far more people available to help. If there is a bug or you would like to make a feature request, please open a new issue and fill out the template.
Thanks!
